### PR TITLE
[develop < T0046-GA] Add order by descending

### PR DESCRIPTION
**Pull request by katarinasupe**
### Description

What's new:
- Exceptions for ordering.
- Enums for Where (WHERE, AND, OR, XOR, NOT) and Order (ASC, ASCENDING, DESC, DESCENDING).
- All where methods (where, and, or, xor) use their classes to construct queries. Those classes inherit from the `WhereConditionPartialQuery` class. 
- All where_not methods (where_not, and_not, or_not, xor_not) use their classes to construct queries. Those classes inherit from the `WhereNotConditionPartialQuery` class, which inherits from the `WhereConditionPartialQuery` class.
- There is one order_by method that allows either property (default ascending order), a tuple of property and order, or a list of either property or tuple of property and order.

### Pull request type

- [x] Feature

######################################

### Reviewer checklist (the reviewer checks this part)
- [ ] Core feature implementation
- [ ] Tests
- [ ] Code documentation
- [ ] Documentation on [memgraph/docs](https://github.com/memgraph/docs)

######################################

### CODE REVIEW

File: gqlalchemy/query_builder.py [hunk 1]
```diff
@@ -26,6 +26,7 @@ class DeclarativeBaseTypes:
     CALL = "CALL"
     CREATE = "CREATE"
     DELETE = "DELETE"
+    DESC = "DESC"
```
Comment (g-despot) [inline gqlalchemy/query_builder.py:29]: Why did you add it here?

File: tests/memgraph/test_query_builder.py [hunk 2]
```diff
@@ -428,6 +428,16 @@ def test_orderby_desc(memgraph):
     mock.assert_called_with(expected_query)
 
 
+def test_order_by_desc(memgraph):
+    query_builder = QueryBuilder().match().node(variable="n").return_().order_by_desc("n.id")
```
Comment (BorisTasevski) [inline tests/memgraph/test_query_builder.py:432]: key for parameter in `order_by_desc` omitted

**Review by BorisTasevski — CHANGES_REQUESTED** [review]

File: gqlalchemy/query_builder.py [hunk 3]
```diff
@@ -336,6 +337,17 @@ def construct_query(self) -> str:
...
 
+class OrderByDescPartialQuery(PartialQuery):
+    def __init__(self, properties: str):
+        super().__init__(DeclarativeBaseTypes.DESC)
+
+        self.properties = properties
+
+    def construct_query(self) -> str:
+        """Creates a ORDER BY statement Cypher partial query."""
+        return OrderByPartialQuery(self.properties).construct_query() + "DESC "
```
[the remark this answers is not in the document]
Comment (katarinasupe) [reply gqlalchemy/query_builder.py:348]: Because it needs to have one space extra at the end, I will try solving that in another way. I'll let you know once I fix it.

[the remark this answers is not in the document]
Comment (BorisTasevski) [reply gqlalchemy/query_builder.py:348]: I feel that most of the code needs to be rewritten.

I would write something like:

`return f"{OrderByPartialQuery(self.properties).construct_query()}{DeclarativeBaseTypes.DESC} "`

File: gqlalchemy/query_builder.py (hunk 1 above)
Comment (katarinasupe) [reply gqlalchemy/query_builder.py:29]: Okay, my fault, totally did this wrong

File: gqlalchemy/query_builder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (katarinasupe) [reply gqlalchemy/query_builder.py:348]: As Ivan noticed above, I was wrong with putting the DESC in DeclarativeBaseTypes. Therefore, we don't have constants for any of the strings in the query. That's why I need to use "DESC ", like "ORDER BY "which was used above in `order_by()` method. But I agree, this will need to be changed sometimes.

[the remark this answers is not in the document]
Comment (katarinasupe) [reply gqlalchemy/query_builder.py:348]: With the current code base, the correct would be:

`return f"{OrderByPartialQuery(self.properties).construct_query()}DESC "`

[the remark this answers is not in the document]
Comment (katarinasupe) [reply gqlalchemy/query_builder.py]: I pushed new commit, let me know if that's fine.

[the remark this answers is not in the document]
Comment (katarinasupe) [reply gqlalchemy/query_builder.py]: I pushed the version with the inheritance, we can discuss it tomorrow. If it's not okay, I can revert it to the original version.

File: tests/memgraph/test_query_builder.py [hunk 4]
```diff
@@ -428,6 +429,63 @@ def test_orderby_desc(memgraph):
...
+
+    with patch.object(Memgraph, "execute_and_fetch", return_value=None) as mock:
+        query_builder.execute()
+
+    mock.assert_called_with(expected_query)
+
+
+def test_order_by_wrong_ordering(memgraph):
+    with pytest.raises(GQLAlchemyMissingOrdering):
+        QueryBuilder().match().node(variable="n").return_().order_by(("n.id", "DESCE")).execute()
```
Comment (katarinasupe) [inline tests/memgraph/test_query_builder.py:444]: It is not clear to me why `GQLAlchemyMissingOrdering` error is being raised only if I add `execute() `at the end of the query builder. I did [a similar thing in `where()`](https://github.com/memgraph/gqlalchemy/pull/111/files#diff-b58f85388321491b6930511b2eae02b960636ea7c5b2d2ab20b465dc735471f6R344) clause and `execute()` wasn't necessary.

Comment (BorisTasevski) [reply tests/memgraph/test_query_builder.py:445]: Because we are doing checks at the time of the execution and not in creation time.

File: gqlalchemy/query_builder.py [hunk 5]
```diff
@@ -609,7 +640,9 @@ def return_(self, results: Optional[Dict[str, str]] = {}) -> "DeclarativeBase":
 
         return self
 
-    def order_by(self, properties: str) -> "DeclarativeBase":
+    def order_by(
+        self, properties: Union[str, Tuple[str, Order], List[Union[str, Tuple[str, Order]]]]
+    ) -> "DeclarativeBase":
```
Comment (brunos252) [inline gqlalchemy/query_builder.py:645]: language question, why do we need qoutes around DeclarativeBase?

File: gqlalchemy/query_builder.py [hunk 6]
```diff
@@ -330,14 +340,35 @@ def construct_query(self) -> str:
...
+    def __init__(self, properties: Union[str, Tuple[str, Order], List[Union[str, Tuple[str, Order]]]]):
         super().__init__(DeclarativeBaseTypes.ORDER_BY)
 
         self.properties = properties
 
     def construct_query(self) -> str:
         """Creates a ORDER BY statement Cypher partial query."""
-        return f" ORDER BY {self.properties} "
+
+        return f" {self.type} {(self._read_list(self.properties)) if isinstance(self.properties, list) else self._read_item(self.properties)} "
```
Comment (brunos252) [inline gqlalchemy/query_builder.py:351]: This line is really long, [PEP8](https://peps.python.org/pep-0008/#maximum-line-length) says it should be up to 80 characters, but in our checks flake8 permits it, Why? Also, should we avoid such long lines for readability?

**Comment by brunos252** [comment]
Do you plan on making more comments in the code?

File: gqlalchemy/query_builder.py (hunk 5 above)
Comment (BorisTasevski) [reply gqlalchemy/query_builder.py:647]: https://peps.python.org/pep-0484/#forward-references

File: gqlalchemy/query_builder.py (hunk 6 above)
Comment (BorisTasevski) [reply gqlalchemy/query_builder.py]: Fixed

**Comment by katarinasupe** [comment]
@brunos252  I think the plan is to document everything better here: https://github.com/memgraph/gqlalchemy/pull/116

File: gqlalchemy/query_builder.py [hunk 7]
```diff
@@ -139,15 +146,46 @@ def construct_query(self) -> str:
...
 
 class WhereConditionPartialQuery(PartialQuery):
-    def __init__(self, keyword: str, query: str):
-        super().__init__(DeclarativeBaseTypes.WHERE)
+    def __init__(self, item: str, operator: str, keyword: str = None, **kwargs):
+        super().__init__(type=DeclarativeBaseTypes.WHERE)
 
         self.keyword = keyword
-        self.query = query
+        self._build_where_query(item=item, operator=operator, **kwargs)
```
Comment (katarinasupe) [inline gqlalchemy/query_builder.py:153]: If this method is called in init and not in construct query, the type checks are done before execution?

File: gqlalchemy/query_builder.py [hunk 8]
```diff
@@ -139,15 +146,46 @@ def construct_query(self) -> str:
...
+                raise GQLAlchemyLiteralAndExpressionMissingInWhere
+
+            value = to_cypher_value(literal)
+        elif literal is not None:
+            raise GQLAlchemyExtraKeywordArgumentsInWhere
+
+        self.query = ("" if operator == ":" else " ").join([item, operator, value])
+
+
+class AndWhereConditionPartialQuery(WhereConditionPartialQuery):
```
Comment (katarinasupe) [inline gqlalchemy/query_builder.py:176]: This is much better 😄

File: gqlalchemy/query_builder.py [hunk 9]
```diff
@@ -374,7 +437,7 @@ def construct_query(self) -> str:
 
 class DeclarativeBase(ABC):
     def __init__(self, connection: Optional[Union[Connection, Memgraph]] = None):
-        self._query: List[Any] = []
+        self._query: List[PartialQuery] = []
```
Comment (katarinasupe) [inline gqlalchemy/query_builder.py:440]: Nice, we obviously missed this.

File: gqlalchemy/query_builder.py [hunk 10]
```diff
@@ -609,9 +658,23 @@ def return_(self, results: Optional[Dict[str, str]] = {}) -> "DeclarativeBase":
...
+        Args:
+            properties: Properties and order by which the query results will be ordered.
+
+        Raises:
+            GQLAlchemyOrderByTypeError: Raises an error when the given ordering is of the wrong type.
+            GQLAlchemyMissingOrdering: Raises an error when the given property is neither string nor tuple.
+
+        Returns:
+            self: A partial Cypher query built from the given parameters.
+        """
```
Comment (g-despot) [inline gqlalchemy/query_builder.py:675]: I would add this additional comment:
```
Examples:
            Ordering query results by the property `n.name` in ascneding order
            and by the property `n.last_name` in descending order:

            Python: `match().node(variable="n").return_().order_by(properties=["n.name", ("n.last_name", Order.DESC)])`
            Cypher: `MATCH (n) RETURN * ORDER BY n.id DESC, n.name, n.last_name DESC;`
```

**Review by g-despot — APPROVED** [review]

File: gqlalchemy/query_builder.py (hunk 7 above)
Comment (BorisTasevski) [reply gqlalchemy/query_builder.py:153]: Yes, the plan is to validate the query at a construction time

**Review by brunos252 — APPROVED** [review]

File: gqlalchemy/query_builder.py [hunk 11]
```diff
@@ -139,15 +147,68 @@ def construct_query(self) -> str:
...
-        self.keyword = keyword
-        self.query = query
+    def __init__(self, item: str, operator: str, keyword: Where = Where.WHERE, is_negated: bool = False, **kwargs):
+        super().__init__(type=keyword.name if not is_negated else f"{keyword.name} {Where.NOT.name}")
+        self.query = self._build_where_query(item=item, operator=operator, **kwargs)
 
     def construct_query(self) -> str:
         """Constructs a where partial query."""
-        return f" {self.keyword} {self.query} "
+
```
Comment (Josipmrden) [inline gqlalchemy/query_builder.py:160]: Delete this line. Or we discuss whether we keep the empty line after docstring.

File: gqlalchemy/query_builder.py [hunk 12]
```diff
@@ -59,11 +59,19 @@ class MatchConstants:
...
+    OR = 3
+    XOR = 4
+    NOT = 5
+
+
+class Order(Enum):
+    ASC = 1
+    ASCENDING = 2
+    DESC = 3
+    DESCENDING = 4
```
Comment (Josipmrden) [inline gqlalchemy/query_builder.py:74]: Do we need 4 of them? Can't we use 2?

File: gqlalchemy/query_builder.py [hunk 13]
```diff
@@ -470,30 +556,50 @@ def from_(
...
+            Filtering query results by the node label.
+
+            Python: `match().node(variable="n").where(item="n", operator=":", expression="User").return_()`
+            Cypher: `MATCH (n) WHERE n:User RETURN *;`
+
+            Filtering query results by the comparison of node property and literal.
+
+            Python: `match().node(variable="n").where(item="n.age", operator=">", literal=18).return_()`
+            Cypher: `MATCH (n) WHERE n.age > 18 RETURN *;`
+        """
```
Comment (Josipmrden) [inline gqlalchemy/query_builder.py:592]: Nice docs :+1:

**Review by BorisTasevski — APPROVED** [review]

**Review by Josipmrden — COMMENTED** [review]
Few comments left

File: gqlalchemy/query_builder.py (hunk 12 above)
Comment (BorisTasevski) [reply gqlalchemy/query_builder.py:74]: All of them are valid

File: gqlalchemy/query_builder.py (hunk 13 above)
Comment (katarinasupe) [reply gqlalchemy/query_builder.py:592]: Thanks 😄  @g-despot made me do it 😂

File: gqlalchemy/query_builder.py (hunk 11 above)
Comment (katarinasupe) [reply gqlalchemy/query_builder.py:160]: There are a couple of occurrences of this, so let's discuss it. I'll put this in the channel.
